In [1]:
!pip install pingouin
import pingouin as pg

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 204.4/204.4 kB 9.4 MB/s eta 0:00:00


In [2]:
import pandas as pd #Library for data manipulation and analysis. In particular, it offers data structures and operations for manipulating numerical tables and time series.

df = pd.read_csv('일원분산분석_실습.csv', sep=',', encoding='CP949')
df

store  satisfaction
0       1             1
1       1             5
2       1             4
3       1             3
4       1             3
5       1             3
6       1             3
7       2             5
8       2             5
9       2             4
10      2             5
11      2             5
12      2             3
13      2             2
14      2             4
15      3             1
16      3             1
17      3             2
18      3             2
19      3             1
20      3             1
21      3             1
22      3             1
23      3             1
24      3             2

In [4]:
import scipy as sp
import numpy as np
import pingouin as pg
df1 = df[['store', 'satisfaction']]
store = []
for i in range(1, 4):
    store.append(df1[df1['store']==i]['satisfaction'].values)
sp.stats.levene(store[0], store[1], store[2])  #일원분산분석 전에 등분산성을 해야함. 여기선 levene의 등분산성 사용. pvalue0.2여서 귀무가설 채택. 차이가 없다=등분산성 이루고 있다.

LeveneResult(statistic=np.float64(1.5399310992796744), pvalue=np.float64(0.23663167821410097))

In [5]:
# One-way ANOVA
print(pg.anova(dv='satisfaction', between='store', data=df1))

  Source  ddof1  ddof2          F     p-unc       np2
0  store      2     22  20.637529  0.000009  0.652312


pvalue 0,00009 대립가설 채택.  store 1,2,3, 만족도 값에 차이가 있다.

등분산성이 만족하지 않을 경우 welch anova라는 코드를 사용

print(pg.welch_anova(dv='satisfaction', between='store', data=df1))

위에는 차이가 있다는 것만 알 뿐, 구체적으로 어떤 집단과 어떤 집단의 만족도에 차이가 있는지를 모름.  그래서 **사후 분석**이라는 것을 해야한다.

In [6]:
!pip install scikit-posthocs

**scikit-posthocs** is a Python library for performing multiple comparisons (post-hoc tests) after statistical tests like ANOVA or Kruskal-Wallis.
It is especially useful when you want to find which specific groups are different from each other.

In [7]:
import scikit_posthocs

df1['store'] = df1['store'].astype(str)
print(scikit_posthocs.posthoc_scheffe(df1, val_col='satisfaction', group_col='store'))

          1         2         3
1  1.000000  0.159548  0.002823
2  0.159548  1.000000  0.000013
3  0.002823  0.000013  1.000000


A와 B store은 차이가 없다. pvalue->0.15
A와 C store은 차이가 있다. pvalue->0.0028
B와 C store이 차이가 있다. pvalue->0.000013